File: day4/part1/securecontainer.py


def validate_passwd(passwd, start, end):
    if passwd > end or passwd < start:
        return False

    lst_passwd = list(str(passwd))

    # Six-digit number
    if len(lst_passwd) != 6:
        return False

    # If decreases
    for i in range(len(lst_passwd)):
        for j in range(i+1, len(lst_passwd)):
            if int(lst_passwd[i]) > int(lst_passwd[j]):
                return False

    # If two adjacent digits are the same
    has_adjacent = False
    for i in range(len(lst_passwd)-1):
        if lst_passwd[i] == lst_passwd[i+1]:
            has_adjacent = True

    if has_adjacent:
        return True
    else:
        return False

if __name__ == "__main__":
    start = 307237
    end = 769058

    valid_passwds = 0
    for passwd in range(start, end+1):
        if validate_passwd(passwd, start, end):
            valid_passwds += 1

    print(valid_passwds)

File: day4/part1/securecontainer.go
package main

import (
	"fmt"
	"strconv"
)

func validate(passwd, start, end int) bool {
	lstPasswd := []rune(strconv.Itoa(passwd))

	// Is a six-digit number
	if len(lstPasswd) != 6 {
		return false
	}

	// Within the range [start, end] inclusive
	if passwd < start || passwd > end {
		return false
	}

	// Digits never decrease
	for i := 0; i < len(lstPasswd)-1; i++ {
		if lstPasswd[i] > lstPasswd[i+1] {
			return false
		}
	}

	// Have two adjacent digits
	adjacent := false
	for i := 0; i < len(lstPasswd)-1; i++ {
		if lstPasswd[i] == lstPasswd[i+1] {
			adjacent = true
		}
	}

	if adjacent {
		return true
	}

	return false
}

func main() {
	start := 307237
	end := 769058

	validPasswds := 0
	for i := start; i <= end; i++ {
		if validate(i, start, end) {
			validPasswds++
		}
	}
	fmt.Println(validPasswds)
}
